internal/providers/vault/certificate.go

```diff
@@ -26,30 +23,23 @@ const (
 	vaultRequestTimeout = 10 * time.Second
 )
 
-func Register() {
-	p := &provider{
-		vaultClients: make(map[string]*vaultclient.Client),
+func (p *provider) certificateProvider(ctx context.Context, secretId secrets.SecretIdentifier, cfg *vault.Certificate) ([]byte, error) {
+	vp := cfg.GetProvider()
+	if vp == nil {
+		return nil, fnerrors.BadInputError("invalid vault certificate configuration: missing provider configuration")
 	}
-	combined.RegisterSecretsProvider(
-		func(ctx context.Context, secretId secrets.SecretIdentifier, cfg *vault.Certificate) ([]byte, error) {
-			ca := cfg.GetCa()
-			if ca == nil {
-				return nil, fnerrors.BadInputError("invalid vault certificate configuration: missing CA configuration")
-			}
 
-			vaultClient, err := p.Login(ctx, ca, vaultJwtAudience)
-			if err != nil {
-				return nil, err
-			}
+	vaultClient, err := p.Login(ctx, vp, vaultJwtAudience)
+	if err != nil {
+		return nil, err
+	}
 
-			if secretId.ServerRef == nil {
-				return nil, fnerrors.BadDataError("required server reference is not set")
-			}
+	if secretId.ServerRef == nil {
+		return nil, fnerrors.BadDataError("required server reference is not set")
+	}
 
-			commonName := fmt.Sprintf("%s.%s", strings.ReplaceAll(secretId.ServerRef.RelPath, "/", "-"), cfg.GetBaseDomain())
-			return p.IssueCertificate(ctx, vaultClient, ca.GetIssuer(), commonName)
-		},
-	)
+	commonName := fmt.Sprintf("%s.%s", strings.ReplaceAll(secretId.ServerRef.RelPath, "/", "-"), cfg.GetBaseDomain())
```

Comment:
I know you're just refactoring but the cname generation here needs revisiting; we don't use server references as stable identifiers as they depend on the location in the codebase.

We'd use server ids for that.

But those hard to digest, so I'd be inclined to receive the cname explicitly as an argument, instead of generating it.

Reply:
I actually added this [here](https://github.com/namespacelabs/foundation/pull/1229)
Initially I was thinking to have less secrets defined in `fn-workspace.cue`, but now I also on the side to provide CN explicitly.
I will fix this.

Reply:
OK I'll leave it as-is for now, but then it would be cool if we maybe _required_ the user to specify the cname (or… default tot he server ID unless one is specified?)

Reply:
server ID could be an option (e.g. the one defined in `server.cue`), but not sure if it easy to get it at this phase (but didn't spend much time trying to make it work).

Reply:
My suggestion would be: keep it as you added already, since it works, and then switch to using server-id in a follow-up change, and then merge your PR that adds custom CN.